asab/api/discovery.py

```diff
@@ -33,81 +54,170 @@ async def locate(self, instance_id: str = None, service_id: str = None) -> list:
 		return [
 			"http://{}:{}".format(servername, port)
 			for servername, port
-			in await self._locate(instance_id, service_id)
+			in await self._locate(instance_id, **kwargs)
 		]
 
-	async def _locate(self, instance_id: str = None, service_id: str = None) -> typing.List[typing.Tuple]:
+	async def _locate(self, instance_id: str = None, **kwargs) -> typing.Set[typing.Tuple]:
 		"""
-		Locates service instances based on their instance ID or service ID.
+		Locates service instances based on their identifiers.
 
 		:param instance_id: The unique identifier for a specific instance of a service
 		:type instance_id: str
-		:param service_id: The ID of the service to locate
-		:type service_id: str
 		:return: a list of tuples containing the server name and port number of the located service(s).
 		"""
-		if instance_id is None and service_id is None:
-			L.warning("Please provide instance_id, service_id, or appclass to locate the service(s).")
-			return
+		res = set()
 
-		instances = await self.get_advertised_instances()
-		if len(instances) == 0:
-			L.warning("No instances available.")
-			return
-		res = []
-		for instance in instances:
-			if service_id is not None:
-				if service_id != instance.get("service_id", "").lower():
-					continue
+		if instance_id is not None:
+			locate_params = {"instance_id": instance_id}
+		elif len(kwargs) > 0:
+			locate_params = kwargs
+		else:
+			L.warning("Please provide instance_id, service_id, or other custom id to locate the service(s).")
+			return res
 
-			if instance_id is not None:
-				if instance_id != instance.get("instance_id", "").lower():
-					continue
+		if self.AdvertisedCache is None:
+			L.warning("Discovery service is not yet ready to locate services.")
+			return res
 
-			web = instance.get("web")
-			host = instance.get("node_id", instance.get("host"))
+		if len(self.AdvertisedCache) == 0:
```

Comment:
No special case, please.

Reply:
This warning indicates that no instances were found. Most probably because of wrong configuration. Make the warning actionable - tell the user to check config